### PR TITLE
improvement: Use inlay hints for worksheets

### PR DESCRIPTION
**Pull request by tgodzik**
inlay-worksheets

Advantages:
- this would allow us to stop using a custom extension and support more editors such as Zed
- can reuse any new enhancement to inlay hints

Disadvantages:
- no way to set color
- could mix with normal decorations

**I plan on removing decoration protocol and worksheet fallback in this PR**

TODO
- [x] remove worksheet publishers
- [x] add specific test to using inlay hints

### CODE REVIEW

**Comment by tgodzik** [comment]
@ckipp01 do you know if inlay hints would work ok for nvim-metals? I want to remove as many custom extensions as I can since we get more and more editors and implementing those custom extension might be too much of a hassle.

**Comment by ckipp01** [comment]
> @ckipp01 do you know if inlay hints would work ok for nvim-metals?

Yup, inlay hints are supported in nvim, so this should work. If I get a bit of time I'll try this pr out with nvim-metals and see if I hit on any issues.

**Comment by tgodzik** [comment]
> > @ckipp01 do you know if inlay hints would work ok for nvim-metals?
> 
> Yup, inlay hints are supported in nvim, so this should work. If I get a bit of time I'll try this pr out with nvim-metals and see if I hit on any issues.

Thanks! I think I mostly need to fix tests here. I will also see how Zed handles it since I managed to get that set up for myself.

**Comment by dragove** [comment]
I did some test on `nvim-metals`. here are my feedbacks.

1. Evaluation results showed using inlay hints style as expected.
2. Adding a character to make code not compilable and the inlay hints not go away (while vscode only shows the error and hindes the inlay hints)
3. Deleting the character to make code compilable again then I can not see the evaluation results.
4. use `:e` to reload the file things get working again.

There is one thing should be mentioned. Using inlay hints also means disabling editor's inlay hints feature will cause results not showing. Example configs are:

1. vscode's `"editor.inlayHints.enabled": "off"`
2. neovim's `vim.lsp.inlay_hint.enable(false)`

**Comment by tgodzik** [comment]
> Adding a character to make code not compilable and the inlay hints not go away (while vscode only shows the error and hindes the inlay hints)

Looks like maybe it's because of lacking inlayHints/refresh ? Does neovim support that?

**Comment by tgodzik** [comment]
> There is one thing should be mentioned. Using inlay hints also means disabling editor's inlay hints feature will cause results not showing. Example configs are:

Sure, but it's enabled by default (at least in VS Code) and other editor users might be sued to turning things on. We need to write about it in the release notes for sure.

File: metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala [hunk 1]
```diff
@@ -160,20 +157,77 @@ class WorksheetProvider(
...
+          .statements()
+          .map { stat =>
+            val statEnd = stat.position().toLsp.getEnd()
+            // Update positions so that they don't break
+            distance.toRevised(statEnd) match {
+              case Left(_) =>
+              case Right(right) =>
+                statEnd.setLine(right.startLine)
+                statEnd.setCharacter(right.startColumn)
+            }
```
Comment (kasiaMarek) [inline metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala:220]:
```suggestion
            distance.toRevised(statEnd).foreach { right =>
                statEnd.setLine(right.startLine)
                statEnd.setCharacter(right.startColumn)
            }
```

File: tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala [hunk 2]
```diff
@@ -283,18 +226,20 @@ abstract class BaseWorksheetLspSuite(
            |""".stripMargin
       )
       _ <- server.didOpen("a/src/main/scala/Main.worksheet.sc")
-      _ <- cancelled.future
-      _ = client.onWorkDoneProgressStart = (_, _) => {}
+      _ <- cancelled.future.withTimeout(10.seconds)
+      _ = client.onWorkDoneProgressStart = (message, _) => {
+        scribe.info(message)
```
Comment (kasiaMarek) [inline tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:231]: ?

File: tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala [hunk 3]
```diff
@@ -266,6 +208,7 @@ abstract class BaseWorksheetLspSuite(
     cleanWorkspace()
     val cancelled = Promise[Unit]()
     client.onWorkDoneProgressStart = { (message, cancelParams) =>
+      scribe.info(message)
```
Comment (kasiaMarek) [inline tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:211]: ?

File: metals/src/main/scala/scala/meta/internal/metals/InlayHintResolveProvider.scala [hunk 4]
```diff
@@ -23,11 +23,19 @@ final class InlayHintResolveProvider(
       token: CancelToken,
   ): Future[InlayHint] = {
     scala.util.Try {
-      val data = inlayHint.getData().asInstanceOf[JsonArray]
-      getLabelParts(inlayHint).zip(parseData(data))
+      Option(inlayHint.getData()) match {
+        case Some(data: JsonArray) =>
+          resolve(
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/InlayHintResolveProvider.scala:28]: I think there was a case where it was throwing and it felt safer.

**Pull request merged**